BJS2-15579 - "Инициативы"

i7rincpane opened the pull request:
1. реализовал метод по созданию инициативы
2. реализовал метод по обновлению инициативы, с обработкой статуса
3. реализовал метод получения иницитив - 1, всех, по фильтру.

Pa4aPv committed:
BJS2-15579 - "Инициативы"
1. реализовал метод по созданию инициативы
2. реализовал метод по обновлению инициативы, с обработкой статуса
3. реализовал метод получения иницитив - 1, всех, по фильтру.

pablo023 commented on src/main/java/faang/school/projectservice/ProjectServiceApplication.java:
```diff
@@ -1,18 +1,33 @@
+            initiativeService.create(new WriteInitiativeDto(4L, "222", "des",
+                    List.of(new WriteStageDto("n1", 4L), new WriteStageDto("n2", 4L)), InitiativeStatus.DONE, 5L));
+        });
+
```
Паш, а что тут происходит?

pablo023 commented on src/main/java/faang/school/projectservice/controller/InitiativeController.java:
```diff
@@ -0,0 +1,37 @@
+@Controller
+@RequiredArgsConstructor
+public class InitiativeController {
+
```
Лишняя пустая строка.

pablo023 commented on src/main/java/faang/school/projectservice/handler/initiative/InitiativeStatusDoneHandler.java:
```diff
@@ -0,0 +1,44 @@
+
+    private Moment createMomentForDoneInitiative(Initiative initiative) {
+        Moment moment = new Moment();
+        moment.setName("Выполненная инициатива");
```
Сюда наверное было бы уместно пихнуть какие-то данные из принимаемой инициативы.

pablo023 commented on src/main/java/faang/school/projectservice/mapper/InitiativeToReadInitiativeDtoMapper.java:
```diff
@@ -0,0 +1,12 @@
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface InitiativeToReadInitiativeDtoMapper {
+
+     ReadInitiativeDto map(Initiative event);
```
Initiative initiative. У нас в проекте есть отдельная сущность эвент. Мелочь конечно но!

pablo023 commented on src/main/java/faang/school/projectservice/mapper/WriteInitiativeDtoToInitiativeMapper.java:
Паш а ты тестируешь это сам?

pablo023 commented on src/main/java/faang/school/projectservice/model/Moment.java:
```diff
@@ -2,14 +2,25 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+@ToString(exclude = {"projects"})
```
Вот это тоже ни к чему. Это логика либо дто, либо сервиса.

pablo023 commented on src/main/java/faang/school/projectservice/model/Moment.java:
```diff
@@ -30,18 +41,20 @@ public class Moment {
             inverseJoinColumns = @JoinColumn(name = "project_id")
     )
-    private List<Project> projects;
+    private List<Project> projects = new ArrayList<>();
```
Не нужно. Мы не используем какие-то конкретные, тем более изменяемы коллекции при декларации сущности.

Pa4aPv committed:
BJS2-15579 - "Инициативы"
1. Обновил контроллер в стиле rest, раписал ендпойнт
2. Добавил глобальный обработчик исключений, добил исключение для отслеживания конфликтов с текущем состоянием ресурса
3. Исправил логику метода поиска по фильтру
4. Добавил логирование везде где вызываются и обрабатываются исключения: в маперы, в валидатор, в сервисные методы и в глобальный обработчик

pablo023 commented on src/main/java/faang/school/projectservice/controller/GlobalExceptionHandler.java:
Ты логируешь стактрейс. Это не всегда далеко нужно, потому что может содержать чувствительные данные,  или потому что слишком много будет логироваться.

pablo023 commented on src/main/java/faang/school/projectservice/controller/ErrorResponse.java:
```diff
@@ -0,0 +1,20 @@
+    private String error;
+    private String message;
+    private String path;
+    private Map<String, String> validationErrors;
```
Зачем здесь это нужно?

pablo023 commented on src/main/java/faang/school/projectservice/controller/ErrorResponse.java:
```diff
@@ -0,0 +1,20 @@
+    private int status;
+    private String error;
+    private String message;
+    private String path;
```
Вот это тоже не особо понятно для чего тут

pablo023 commented on src/main/java/faang/school/projectservice/controller/ErrorResponse.java:
```diff
@@ -0,0 +1,20 @@
+
+    @Builder.Default
+    private final LocalDateTime timestamp = LocalDateTime.now();
+    private int status;
```
А это зачем?

pablo023 commented on src/main/java/faang/school/projectservice/controller/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,79 @@
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ErrorResponse handleEntityIllegalArgument(IllegalArgumentException ex) {
+        log.error("IllegalArgumentException", ex);
+        return ErrorResponse.builder()
+                .status(HttpStatus.BAD_REQUEST.value())
+                .error(HttpStatus.BAD_REQUEST.getReasonPhrase())
+                .message(ex.getMessage())
+                .build();
+    }
```
Вот этот метод - практически одно и то же в каждом методе пишешь. Вот это бы выделить в отдельный метод.

pablo023 commented on src/main/java/faang/school/projectservice/controller/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,79 @@
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ErrorResponse handleValidationExceptions(MethodArgumentNotValidException ex) {
+        log.error("MethodArgumentNotValidException", ex);
+        Map<String, String> errors = new HashMap<>();
+        ex.getBindingResult().getFieldErrors().forEach(error -> {
+            errors.put(error.getField(), error.getDefaultMessage());
+        });
```
Какой в этом смысл? Ты просто залогируешь стак трейс полностью, никаким образом не добавив подсказок - что произошло. Ты просто то, что в консоль пишется, напишешь в текстовый файл. В таком логировании смысла нет. Логировать имеет смысл так, чтобы легко было понятно - где, что, и почему легло.

Сервис, метод, ошибка, входные данные - вот примерно так.

pablo023 commented on src/main/java/faang/school/projectservice/dto/client/initiative/ReadInitiativeDto.java:
```diff
@@ -0,0 +1,18 @@
+import lombok.Value;
+
+
+@Value
```
Я не очень понимаю зачем вот это делать, чтобы не писать просто модификаторы доступа?

pablo023 commented on src/main/java/faang/school/projectservice/dto/client/initiative/WriteInitiativeDto.java:
Давай разберемся, почему использовать аннотацию Value в твоих DTO — это не самый крутой вариант.

Когда ты лепишь аннотацию Value, твой класс становится иммутабельным. Lombok делает следующие вещи:

- все поля становятся private final
- генерируется конструктор, который принимает все поля как параметры
- генерятся геттеры для всех полей.
- equals(), hashCode() и toString() методы тоже добавляются.

Вроде бы круто, но для DTO это не всегда подходит. DTO — это такие болванчики для передачи данных между слоями приложения (например, от клиента к серверу). И часто нужно, чтобы они были изменяемыми, особенно когда мы юзаем фреймворки для сериализации, такие как Jackson. Они обычно полагаются на конструкторы без параметров и сеттеры, чтобы десериализовать JSON в Java-объекты. А с @Value у нас нет конструктора по умолчанию и все поля final. Это проблема. Jackson не сможет создать экземпляр этого класса из JSON без специального десериализатора. А это уже лишние заморочки. Да, Value — норм штука для сокращения шаблонного кода в value-объектах, где важна иммутабельность. Но использовать её в DTO — это перебор. DTO должны быть простыми переносчиками данных с минимальной логикой. Лучше использовать Data или просто явно определить геттеры, сеттеры и, возможно, конструкторы.

pablo023 commented on src/main/java/faang/school/projectservice/valitator/initiative/InitiativeStatusDoneValidator.java:
Создавать класс под один метод - это перебор

pablo023 replied:
Это примерно как пульт для телевизора, который может включить только пятый канал.

pablo023 commented on src/main/java/faang/school/projectservice/valitator/initiative/WriteInitiativeDtoValidator.java:
```diff
@@ -0,0 +1,71 @@
+
+    private void checkDescriptionIsNotNull(WriteInitiativeDto writeInitiativeDto) {
+        if (writeInitiativeDto.getDescription() == null) {
+            log.error("WriteInitiativeDtoValidator.checkDescriptionIsNotNull: Description cannot be null");
```
Не особо имеет смысл, хоть и неплохой подход. Ты если вот это в логах увидишь - понятно тебе не станет, в каком сервисе, и где конкретно это отлетело. Валидатор вот конкретно вот этот может использоваться в разных методах, и даже в разных сервисах. Короче, я бы в сервисе это логировал, но не суть прям важно

pablo023 commented on src/main/java/faang/school/projectservice/service/InitiativeService.java:
```diff
@@ -0,0 +1,88 @@
+        return initiativeRepository.findById(id)
+                .map(entity -> {
+                    initiativeHandlers.stream()
+                            .filter(handler -> handler.isApplicable(writeInitiativeDto, entity))
```
какой хэндлер, что хэндл.. Не должно быть так, что я смотрю на название объекта, и его метод, и даже догадаться не могу, что он и для чего делает, а мне надо идти и смотреть код

pablo023 commented on src/main/java/faang/school/projectservice/service/InitiativeService.java:
```diff
@@ -0,0 +1,88 @@
+                            .forEach(handler -> handler.handle(entity));
+                    return entity;
+                })
+                .map(entity -> writeInitiativeDtoToInitiativeMapper.map(writeInitiativeDto, entity))
```
мап куда? Я из названий твоих методов не понимаю, что я получу от них

pablo023 commented on src/main/java/faang/school/projectservice/model/Project.java:
```diff
@@ -82,9 +89,10 @@ public class Project {
+    @Builder.Default
     @ManyToMany(mappedBy = "projects")
-    private List<Moment> moments;
+    private List<Moment> moments = new ArrayList<>();
```
Не надо так.

Когда мы работаем с Hibernate, важно помнить о некоторых особенностях инициализации коллекций в наших сущностях. Hibernate предпочитает самостоятельно управлять коллекциями в классах, используя свои реализации для поддержки ленивой загрузки и отслеживания изменений. Это означает, что прямая инициализация коллекций в полях класса (например, private List<Moment> moments = new ArrayList<>();) может привести к конфликтам. Вместо инициализации коллекций в месте объявления лучше выполнить это в конструкторе по умолчанию, если это ПРЯМ ВОТ НАДО. Скорее всего - не надо (тебе вот тут это не надо). Но если и надо, то так, это позволяет Hibernate при необходимости заменить эти коллекции своими реализациями при загрузке сущности из базы данных.

pablo023 reviewed (changes requested):
Много лишнего пишешь, не понимаю зачем, если честно. Не нужно реализовывать требования, которых нет. Возвращаю в основном из-за 51 строки в Moment. В остальном - просто много лишнего, и неправильно логируешь

pablo023 commented:
Исправь момент, тестируй, фикси конфликты, и мерж
